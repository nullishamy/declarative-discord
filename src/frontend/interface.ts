import { GuildConfiguration } from "../util/schema.js";

interface ParseResultOk {
  success: true;
  data: GuildConfiguration;
}

interface ParseResultErr {
  success: false;
  err: Error;
}

export type ParseResult = ParseResultOk | ParseResultErr;

export interface Frontend {
  name: string;

  parseFromData(data: string): Promise<ParseResult>;
  parseFromFile(path: string): Promise<ParseResult>;
}
