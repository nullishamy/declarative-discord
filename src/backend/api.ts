import { REST } from '@discordjs/rest'
import assert from 'assert'
import {
  APIChannel,
  APIGuild,
  APIOverwrite,
  APIRole,
  ChannelType,
  Routes
} from 'discord-api-types/v10'
import { Category, GuildChannelWithOpts, Role, RoleOverride } from '../util/schema.js'
import {
  AllDisabledPerms,
  bitfieldToString,
  stringToBitField
} from './permissions.js'

export abstract class API {
  public static VERSION = '10'

  protected readonly rest: REST
  constructor (protected readonly guildId: string, token: string) {
    this.rest = new REST({ version: API.VERSION }).setToken(token)
  }

  abstract fetchGlobalChannels (): Promise<GuildChannelWithOpts[]>
  abstract fetchRoles (): Promise<Role[]>
  abstract fetchCategories (): Promise<Category[]>

  abstract pushChannel (channel: GuildChannelWithOpts): Promise<void>
  abstract pushCategory (category: Category): Promise<void>
  abstract pushRole (role: Role): Promise<void>
}

export class APIImpl extends API {
  private fetchedGuild: APIGuild | undefined
  private fetchedChannels: APIChannel[] | undefined

  private mappedRoles: Record<string, APIRole> | undefined
  private mappedChannels: Record<string, APIChannel> | undefined

  async fetchGlobalChannels (): Promise<GuildChannelWithOpts[]> {
    const channels: APIChannel[] =
      this.fetchedChannels ??
      ((await this.rest.get(
        Routes.guildChannels(this.guildId)
      )) as APIChannel[])
    this.fetchedChannels = channels
    await this.fetchRoles()

    this.applyMappings()

    assert(this.mappedRoles)

    const mappedRoles = this.mappedRoles

    return this.fetchedChannels
      .filter(
        (c) =>
          (c.type === ChannelType.GuildText ||
            c.type === ChannelType.GuildVoice) &&
          c.parent_id === null
      )
      .map((c) => {
        if (c.type === ChannelType.GuildText) {
          return {
            type: 'text' as const,
            comment: c.name,
            id: c.id,
            parentId: undefined,
            options: {
              nsfw: c.nsfw ?? false,
              slowmode: c.rate_limit_per_user ?? 0,
              topic: c.topic ?? undefined
            },
            overrides: this.roleOverwritesToOverrides(
              c.permission_overwrites ?? [],
              mappedRoles
            )
          }
        }

        if (c.type === ChannelType.GuildVoice) {
          return {
            type: 'voice' as const,
            comment: c.name,
            id: c.id,
            parentId: undefined,
            options: {
              nsfw: c.nsfw ?? false,
              bitrate: c.bitrate,
              userLimit: c.user_limit
            },
            overrides: this.roleOverwritesToOverrides(
              c.permission_overwrites ?? [],
              mappedRoles
            )
          }
        }

        assert(false, 'impossible')
      })
  }

  async fetchRoles (): Promise<Role[]> {
    const guild =
      this.fetchedGuild ??
      ((await this.rest.get(Routes.guild(this.guildId))) as APIGuild)
    this.fetchedGuild = guild
    this.applyMappings()

    return (
      this.fetchedGuild.roles
        // Do not include managed roles, they are controlled by other bots
        .filter((r) => !r.managed)
        .map((r) => {
          // Disable all perms by default
          const permissions = { ...AllDisabledPerms }

          for (const truePerm of bitfieldToString(Number(r.permissions))) {
            // .. enable the ones specified
            permissions[truePerm.toLowerCase()] = true
          }

          return {
            comment: r.name,
            id: r.id,
            permissions
          }
        })
    )
  }

  async fetchCategories (): Promise<Category[]> {
    await this.fetchGlobalChannels()
    await this.fetchRoles()
    this.applyMappings()

    assert(this.fetchedChannels)
    assert(this.mappedRoles)
    assert(this.mappedChannels)

    const { mappedRoles, mappedChannels } = this

    // It is easier to group up the channels with a mapping, but we do not need the mapping after the fact
    return Object.values(
      this.fetchedChannels.reduce<Record<string, Category>>((acc, val) => {
        if (
          val.type === ChannelType.GuildText ||
          val.type === ChannelType.GuildVoice
        ) {
          if (!val.parent_id) {
            return acc
          }

          if (!(val.parent_id in acc)) {
            const parent = mappedChannels[val.parent_id]
            if (parent.type !== ChannelType.GuildCategory) {
              throw new Error(`expected ChannelType.GuildCategory, got ${ChannelType[parent.type]}`)
            }

            // Role overrides for the category
            const roleOverrides: RoleOverride[] =
              this.roleOverwritesToOverrides(
                parent.permission_overwrites ?? [],
                mappedRoles
              )

            acc[val.parent_id] = {
              id: val.parent_id,
              comment: parent.name,
              channels: [],
              overrides: roleOverrides
            }
          }

          if (val.type === ChannelType.GuildText) {
            acc[val.parent_id].channels.push({
              type: 'text' as const,
              comment: val.name,
              id: val.id,
              parentId: val.parent_id,
              options: {
                nsfw: val.nsfw ?? false,
                slowmode: val.rate_limit_per_user ?? 0,
                topic: val.topic ?? undefined
              },
              overrides: this.roleOverwritesToOverrides(
                val.permission_overwrites ?? [],
                mappedRoles
              )
            })
          }

          if (val.type === ChannelType.GuildVoice) {
            acc[val.parent_id].channels.push({
              type: 'voice' as const,
              comment: val.name,
              id: val.id,
              parentId: val.parent_id,
              options: {
                nsfw: val.nsfw ?? false,
                bitrate: val.bitrate,
                userLimit: val.user_limit
              },
              overrides: this.roleOverwritesToOverrides(
                val.permission_overwrites ?? [],
                mappedRoles
              )
            })
          }
        }
        return acc
      }, {})
    )
  }

  async pushChannel (channel: GuildChannelWithOpts): Promise<void> {
    const type = channel.type
    let body: Record<string, unknown> = {}

    if (type === 'voice') {
      body = {
        name: channel.comment,
        nsfw: channel.options.nsfw,
        bitrate: channel.options.bitrate,
        user_limit: channel.options.userLimit
      }
    } else {
      body = {
        name: channel.comment,
        nsfw: channel.options.nsfw,
        topic: channel.options.topic,
        slowmode: channel.options.slowmode
      }
    }

    body.permission_overwrites = channel.overrides.map((r) => {
      let allow = 0n
      let deny = 0n

      for (const [perm, enabled] of Object.entries(r.permissions)) {
        if (enabled) {
          allow = allow | BigInt(stringToBitField(perm.toUpperCase()))
        } else {
          deny = deny | BigInt(stringToBitField(perm.toUpperCase()))
        }
      }

      return {
        id: r.id,
        type: 0, // Always role overrides,
        allow: allow.toString(),
        deny: deny.toString()
      }
    })

    await this.rest.patch(Routes.channel(channel.id), {
      body
    })
  }

  async pushCategory (category: Category): Promise<void> {
    const body = {
      name: category.comment,
      permission_overwrites: category.overrides.map((r) => {
        let allow = 0n
        let deny = 0n

        for (const [perm, enabled] of Object.entries(r.permissions)) {
          if (enabled) {
            allow = allow | BigInt(stringToBitField(perm.toUpperCase()))
          } else {
            deny = deny | BigInt(stringToBitField(perm.toUpperCase()))
          }
        }

        return {
          id: r.id,
          type: 0, // Always role overrides,
          allow: allow.toString(),
          deny: deny.toString()
        }
      })
    }

    await this.rest.patch(Routes.channel(category.id), {
      body
    })
  }

  async pushRole (role: Role): Promise<void> {
    let permissions = 0n

    for (const [perm, enabled] of Object.entries(role.permissions)) {
      // Do not set the bit if the perm is not true (Discord requires us to publish the whole perm, not just a diff)
      if (enabled) {
        const bits = BigInt(stringToBitField(perm.toUpperCase()))
        permissions = permissions | bits
      }
    }

    await this.rest.patch(Routes.guildRole(this.guildId, role.id), {
      body: {
        name: role.comment,
        permissions: permissions.toString()
      }
    })
  }

  private roleOverwritesToOverrides (
    overwrites: APIOverwrite[],
    roleMapping: Record<string, APIRole>
  ): RoleOverride[] {
    return overwrites
      .map((p) => {
        const allowed = bitfieldToString(Number(p.allow)).reduce<Record<string, true>>((acc, val) => {
          acc[val.toLowerCase()] = true
          return acc
        }, {})

        const denied = bitfieldToString(Number(p.deny)).reduce<Record<string, false>>((acc, val) => {
          acc[val.toLowerCase()] = false
          return acc
        }, {})

        return {
          id: p.id,
          comment: roleMapping[p.id].name,
          permissions: { ...allowed, ...denied }
        }
      })
      .sort((a, b) => a.id.localeCompare(b.id))
  }

  private applyMappings () {
    assert(this.fetchedChannels)
    assert(this.fetchedGuild)

    this.mappedChannels = this.fetchedChannels.reduce<Record<string, APIChannel>>((acc, val) => {
      acc[val.id] = val
      return acc
    }, {})

    this.mappedRoles = this.fetchedGuild.roles.reduce<Record<string, APIRole>>((acc, val) => {
      acc[val.id] = val
      return acc
    }, {})
  }
}
