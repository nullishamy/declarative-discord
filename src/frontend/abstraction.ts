import assert from "assert";
import { GuildConfiguration } from "../util/schema.js";

interface ParseResultOk {
  success: true;
  data: GuildConfiguration;
}

interface ParseResultErr {
  success: false;
  err: Error;
}

export type ParseResult = ParseResultOk | ParseResultErr;

export abstract class Frontend {
  protected loadedConfig: GuildConfiguration | undefined;

  constructor(loadedConfig?: GuildConfiguration) {
    this.loadedConfig = loadedConfig;
  }

  assertLoadedConfig() {
    assert(this.loadedConfig, "config not loaded");
    return this.loadedConfig;
  }

  abstract parseFromData(data: string, force: boolean): Promise<ParseResult>;
  abstract parseFromFile(path: string, force: boolean): Promise<ParseResult>;
}
