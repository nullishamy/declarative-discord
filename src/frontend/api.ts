import assert from "assert";
import { LuaFactory } from "wasmoon";
import { z } from "zod";
import { luaLib } from "../runtime/index.js";
import { validated } from "../util/lua.js";
import {
  Category,
  GuildChannelWithOpts,
  GuildConfiguration,
  Role,
  TextChannelWithOpts,
  VoiceChannelWithOpts,
} from "../util/schema.js";

const factory = new LuaFactory();
const engine = await factory.createEngine();

export class GuildSetup {
  public readonly globalChannels: GuildChannelWithOpts[] = [];
  public readonly globalRoles: Role[] = [];
  public readonly categories: Category[] = [];

  constructor(public readonly id: string) {}

  // Orphan setup
  global = {
    text: validated((tbl) => {
      this.globalChannels.push(tbl);
    }, TextChannelWithOpts),
    voice: validated((tbl) => {
      this.globalChannels.push(tbl);
    }, VoiceChannelWithOpts),
    role: validated((tbl) => this.globalRoles.push(tbl), Role),
  };

  // Category setup
  channel = {
    text: (tbl: unknown) => ({
      type: "text",
      ...z.object({}).passthrough().parse(tbl),
    }),
    voice: (tbl: unknown) => ({
      type: "voice",
      ...z.object({}).passthrough().parse(tbl),
    }),
  };

  role = (tbl: unknown) => tbl;

  category = validated((tbl) => this.categories.push(tbl), Category);
}

export class GuildBuilder {
  static LIB_NAME = "discord";

  constructor(private readonly config: string) {}

  async evaluateConfiguration(): Promise<GuildConfiguration> {
    // HACK: maybe find a better way to do this
    // We hack into require to make our lib a module
    const _require: unknown = engine.global.get("require");
    assert(typeof _require === "function", "require was not a function");

    engine.global.set("require", (maybeLibName: unknown) => {
      if (maybeLibName === GuildBuilder.LIB_NAME) {
        return luaLib;
      } else {
        return _require(maybeLibName) as unknown;
      }
    });

    const result: unknown = await engine.doString(this.config);

    // Call the provided setup function after validating the shape
    const validResult = z
      .object({
        id: z.string().nonempty(),
        setup: z.function(),
      })
      .parse(result);

    const setup = new GuildSetup(validResult.id);
    validResult.setup(setup);

    return {
      guildId: setup.id,
      globalChannels: setup.globalChannels.sort((a, b) =>
        a.id.localeCompare(b.id)
      ),
      globalRoles: setup.globalRoles.sort((a, b) => a.id.localeCompare(b.id)),
      categories: setup.categories,
    };
  }
}
