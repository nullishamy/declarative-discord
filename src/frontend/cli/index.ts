import { wrapCommand } from '../../index.js'
import yargs from 'yargs'
import diff from './commands/diff.js'
import lint from './commands/lint.js'
import pull from './commands/pull.js'
import push from './commands/push.js'
import { Args } from './interface.js'
import dump from './commands/dump.js'

let parsed: Args | undefined

export const parseArgs = async () => {
  if (parsed) return parsed

  const defaultBuilder = () => {
    return
  }

  const result = await yargs(process.argv.slice(2))
    .scriptName('dcd')
    .option('config', {
      type: 'string',
      alias: 'c',
      description: 'the config file to use'
    })
    .option('token', {
      type: 'string',
      alias: 't',
      description: 'the token to use, takes priority over the config file'
    })
    .option('discord-config', {
      type: 'string',
      alias: 'd',
      description: 'the discord config to use'
    })
    .option('verbosity', {
      type: 'count',
      alias: 'v',
      description: 'controls the amount of log messages the app produces',
      default: 0
    })
    .command(
      lint.command,
      lint.describe,
      defaultBuilder,
      wrapCommand(lint.handler)
    )
    .command(
      pull.command,
      pull.describe,
      defaultBuilder,
      wrapCommand(pull.handler)
    )
    .command(
      push.command,
      push.describe,
      defaultBuilder,
      wrapCommand(push.handler)
    )
    .command(
      diff.command,
      diff.describe,
      defaultBuilder,
      wrapCommand(diff.handler)
    )
    .command(
      dump.command,
      dump.describe,
      defaultBuilder,
      wrapCommand(dump.handler)
    )
    .help()
    .demandCommand()
    .wrap(null)
    .parse()

  parsed = result
  return result
}
