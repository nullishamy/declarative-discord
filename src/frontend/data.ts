export type OptionValue = string | number | boolean

export type EntityType = 'role' | 'channel' | 'category'

export type Role = {
    id: string,
    comment: string,
    type: 'role',
    permissions: Record<string, boolean | undefined>
}

export type RoleOverride = {
    id: string,
    comment: string,
    type: 'role',
    permissions: Record<string, boolean | undefined>
}


export type VoiceChannel = {
    id: string,
    comment: string,
    type: 'channel',
    channel_type: 'voice'
    options: {
        nsfw: boolean,
        bitrate: number | undefined,
        userLimit: number | undefined
    }
    overrides: RoleOverride[]
}
export type TextChannel = {
    id: string,
    comment: string,
    type: 'channel',
    channel_type: 'text'
    options: {
        nsfw: boolean,
        slowmode: number,
        topic: string | undefined,
    }
    overrides: RoleOverride[]
}

export type Channel = TextChannel | VoiceChannel

export type Category = {
    id: string,
    comment: string,
    type: 'category',
    channels: Channel[],
    overrides: RoleOverride[]
}

export type Entity = Role | Channel | Category

export type GuildConfiguration = {
    id: string
    globalRoles: Role[],
    globalChannels: Channel[],
    categories: Category[]
}
