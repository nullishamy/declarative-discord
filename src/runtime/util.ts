type ConfigVisitorScope = "roles" | "channels";
interface ConfigVisitor {
  scope: ConfigVisitorScope;
  callback: (...args: unknown[]) => unknown;
}

const LibState = {
  // `object` because we need to preserve the input format 1-1
  storedRoles: new Map<string, object>(),
  visitors: new Map<ConfigVisitorScope, ConfigVisitor>(),
};

// Primarily used for testing
export const resetLib = () => {
  LibState.storedRoles = new Map();
  LibState.visitors = new Map();
};

export const wrapLib = (
  fn: (state: typeof LibState, args: unknown[]) => unknown
) => {
  return (...args: unknown[]) => {
    return fn(LibState, args);
  };
};
