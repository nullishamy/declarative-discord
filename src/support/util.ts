import { ZodError } from "zod";
import { prettyZodError } from "../util/error.js";

type ConfigVisitorScope = "roles" | "channels";
interface ConfigVisitor {
  scope: ConfigVisitorScope;
  callback: (...args: unknown[]) => unknown;
}

const LibState = {
  // `object` because we need to preserve the input format 1-1
  storedRoles: new Map<string, object>(),
  visitors: new Map<ConfigVisitorScope, ConfigVisitor>(),
};

// Primarily used for testing
export const resetLib = () => {
  LibState.storedRoles = new Map();
  LibState.visitors = new Map();
};

export const wrapLib = (
  fn: (state: typeof LibState, args: unknown[]) => unknown
) => {
  return (...args: unknown[]) => {
    try {
      return fn(LibState, args);
    } catch (err) {
      if (err instanceof ZodError) {
        logger.error(prettyZodError(err));
      } else {
        logger.error(`Unexpected error:\n${err}`);
      }

      throw err;
    }
  };
};
