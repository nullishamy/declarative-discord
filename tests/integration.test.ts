/* eslint-disable @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment */

import { GuildBuilder } from "../src/frontend/api";
import fs from "fs/promises";
import { resetLib } from "../src/runtime/util.js";

async function loadTestFile(path: string): Promise<GuildBuilder> {
  return new GuildBuilder(await fs.readFile(`./samples/${path}.lua`, "utf-8"));
}

async function runTest(path: string) {
  const builder = await loadTestFile(path);
  const config = await builder.evaluateConfiguration();
  expect(config).toMatchSnapshot();
}

describe("Integrations", () => {
  beforeEach(() => resetLib());

  it("runs minimal.lua", async () => {
    return await runTest("minimal");
  });

  it("runs lib-minimal.lua", async () => {
    return await runTest("lib-minimal");
  });

  it("runs roles.lua", async () => {
    return await runTest("roles");
  });

  it("runs channels.lua", async () => {
    return await runTest("channels");
  });

  it("runs channels-with-overrides.lua", async () => {
    return await runTest("channels-with-overrides");
  });

  it("runs categories.lua", async () => {
    return await runTest("categories");
  });

  // Lib
  it("runs lib-roles.lua", async () => {
    return await runTest("lib-roles");
  });

  it("runs lib-visit.lua", async () => {
    return await runTest("lib-visit");
  });
});
