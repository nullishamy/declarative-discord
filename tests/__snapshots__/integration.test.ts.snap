// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integrations runs categories.lua 1`] = `
{
  "categories": [
    {
      "channels": [
        {
          "comment": "a-channel",
          "id": "1094801227080015882",
          "options": {
            "nsfw": false,
            "slowmode": 0,
            "topic": undefined,
          },
          "overrides": [],
          "parentId": "1094801227080015882",
          "type": "text",
        },
        {
          "comment": "a-voice-channel",
          "id": "1094801227080015882",
          "options": {
            "bitrate": undefined,
            "nsfw": false,
            "userLimit": undefined,
          },
          "overrides": [],
          "parentId": "1094801227080015882",
          "type": "voice",
        },
      ],
      "comment": "a-category",
      "id": "1094801227080015882",
      "overrides": [],
    },
    {
      "channels": [
        {
          "comment": "an-nsfw-channel",
          "id": "1094801227080015882",
          "options": {
            "nsfw": true,
            "slowmode": 0,
            "topic": undefined,
          },
          "overrides": [],
          "parentId": "1094801227080015882",
          "type": "text",
        },
        {
          "comment": "a-voice-channel",
          "id": "1094801227080015882",
          "options": {
            "bitrate": undefined,
            "nsfw": false,
            "userLimit": undefined,
          },
          "overrides": [],
          "parentId": "1094801227080015882",
          "type": "voice",
        },
      ],
      "comment": "another-category",
      "id": "1094801227080015882",
      "overrides": [],
    },
  ],
  "globalChannels": [],
  "globalRoles": [],
  "guildId": "1094801227080015882",
}
`;

exports[`Integrations runs channels.lua 1`] = `
{
  "categories": [],
  "globalChannels": [
    {
      "comment": "test-channel",
      "id": "1094801227080015882",
      "options": {
        "nsfw": false,
        "slowmode": 0,
        "topic": "a topic",
      },
      "overrides": [],
      "parentId": undefined,
      "type": "text",
    },
    {
      "comment": "test-channel-2",
      "id": "1094801227080015882",
      "options": {
        "bitrate": undefined,
        "nsfw": true,
        "userLimit": undefined,
      },
      "overrides": [],
      "parentId": undefined,
      "type": "voice",
    },
    {
      "comment": "test-channel-2",
      "id": "1094801227080015882",
      "options": {
        "nsfw": true,
        "slowmode": 25,
        "topic": undefined,
      },
      "overrides": [],
      "parentId": undefined,
      "type": "text",
    },
    {
      "comment": "test-channel-3",
      "id": "1094801227080015882",
      "options": {
        "nsfw": false,
        "slowmode": 25,
        "topic": undefined,
      },
      "overrides": [],
      "parentId": undefined,
      "type": "text",
    },
  ],
  "globalRoles": [],
  "guildId": "1094801227080015882",
}
`;

exports[`Integrations runs channels-with-overrides.lua 1`] = `
{
  "categories": [],
  "globalChannels": [
    {
      "comment": "test-channel",
      "id": "1094801227080015882",
      "options": {
        "nsfw": false,
        "slowmode": 0,
        "topic": "a topic",
      },
      "overrides": [
        {
          "comment": "mods",
          "id": "1094801227080015882",
          "permissions": {
            "use_vad": true,
          },
        },
        {
          "comment": "mods-2",
          "id": "1094801227080015882",
          "permissions": {
            "send_messages": true,
          },
        },
      ],
      "parentId": undefined,
      "type": "text",
    },
    {
      "comment": "test-channel-2",
      "id": "1094801227080015882",
      "options": {
        "bitrate": undefined,
        "nsfw": true,
        "userLimit": undefined,
      },
      "overrides": [
        {
          "comment": "mods",
          "id": "1094801227080015882",
          "permissions": {
            "speak": false,
          },
        },
        {
          "comment": "mods-2",
          "id": "1094801227080015882",
          "permissions": {
            "priority_speaker": true,
          },
        },
      ],
      "parentId": undefined,
      "type": "voice",
    },
    {
      "comment": "test-channel-2",
      "id": "1094801227080015882",
      "options": {
        "nsfw": true,
        "slowmode": 25,
        "topic": undefined,
      },
      "overrides": [
        {
          "comment": "mods",
          "id": "1094801227080015882",
          "permissions": {
            "manage_messages": true,
          },
        },
        {
          "comment": "mods-2",
          "id": "1094801227080015882",
          "permissions": {
            "attach_files": false,
          },
        },
      ],
      "parentId": undefined,
      "type": "text",
    },
    {
      "comment": "test-channel-3",
      "id": "1094801227080015882",
      "options": {
        "nsfw": false,
        "slowmode": 25,
        "topic": undefined,
      },
      "overrides": [
        {
          "comment": "mods",
          "id": "1094801227080015882",
          "permissions": {
            "manage_roles": true,
          },
        },
        {
          "comment": "mods-2",
          "id": "1094801227080015882",
          "permissions": {
            "move_members": true,
          },
        },
      ],
      "parentId": undefined,
      "type": "text",
    },
  ],
  "globalRoles": [],
  "guildId": "1094801227080015882",
}
`;

exports[`Integrations runs lib-minimal.lua 1`] = `
{
  "categories": [],
  "globalChannels": [],
  "globalRoles": [],
  "guildId": "1094801227080015882",
}
`;

exports[`Integrations runs lib-roles.lua 1`] = `
{
  "categories": [],
  "globalChannels": [
    {
      "comment": "test-channel",
      "id": "1094801227080015882",
      "options": {
        "nsfw": false,
        "slowmode": 0,
        "topic": "a topic",
      },
      "overrides": [
        {
          "comment": "mods-2",
          "id": "42394823948909090",
          "permissions": {
            "move_members": true,
          },
        },
      ],
      "parentId": undefined,
      "type": "text",
    },
  ],
  "globalRoles": [],
  "guildId": "1094801227080015882",
}
`;

exports[`Integrations runs lib-visit.lua 1`] = `
{
  "categories": [],
  "globalChannels": [],
  "globalRoles": [
    {
      "comment": "nil",
      "id": "1094801227080015882",
      "permissions": {
        "add_reactions": false,
        "administrator": false,
        "attach_files": false,
        "ban_members": false,
        "change_nickname": false,
        "connect": false,
        "create_instant_invite": false,
        "create_private_threads": false,
        "create_public_threads": false,
        "deafen_members": false,
        "embed_links": false,
        "kick_members": false,
        "manage_channels": true,
        "manage_events": false,
        "manage_guild": false,
        "manage_guild_expressions": false,
        "manage_messages": false,
        "manage_nicknames": false,
        "manage_roles": false,
        "manage_threads": false,
        "manage_webhooks": false,
        "mention_everyone": false,
        "moderate_members": false,
        "move_members": false,
        "mute_members": false,
        "priority_speaker": false,
        "read_message_history": false,
        "request_to_speak": false,
        "send_messages": false,
        "send_messages_in_threads": false,
        "send_tts_messages": false,
        "speak": false,
        "stream": false,
        "use_application_commands": false,
        "use_embedded_activities": false,
        "use_external_emojis": false,
        "use_external_stickers": false,
        "use_soundboard": false,
        "use_vad": false,
        "view_audit_log": false,
        "view_channel": false,
        "view_creator_monetization_analytics": false,
        "view_guild_insights": false,
      },
    },
  ],
  "guildId": "1094801227080015882",
}
`;

exports[`Integrations runs lint-fails.lua 1`] = `
{
  "categories": [],
  "globalChannels": [
    {
      "comment": "test-channel",
      "id": "1094801227080015882",
      "options": {
        "nsfw": false,
        "slowmode": 0,
        "topic": "",
      },
      "overrides": [],
      "parentId": undefined,
      "type": "text",
    },
  ],
  "globalRoles": [
    {
      "comment": "a-role",
      "id": "1094801227080015882",
      "permissions": {
        "add_reactions": false,
        "administrator": true,
        "attach_files": false,
        "ban_members": false,
        "change_nickname": false,
        "connect": false,
        "create_instant_invite": false,
        "create_private_threads": false,
        "create_public_threads": false,
        "deafen_members": false,
        "embed_links": false,
        "kick_members": false,
        "manage_channels": false,
        "manage_events": false,
        "manage_guild": false,
        "manage_guild_expressions": false,
        "manage_messages": false,
        "manage_nicknames": false,
        "manage_roles": false,
        "manage_threads": false,
        "manage_webhooks": false,
        "mention_everyone": false,
        "moderate_members": false,
        "move_members": false,
        "mute_members": false,
        "priority_speaker": false,
        "read_message_history": false,
        "request_to_speak": false,
        "send_messages": false,
        "send_messages_in_threads": false,
        "send_tts_messages": false,
        "speak": false,
        "stream": false,
        "use_application_commands": false,
        "use_embedded_activities": false,
        "use_external_emojis": false,
        "use_external_stickers": false,
        "use_soundboard": false,
        "use_vad": false,
        "view_audit_log": false,
        "view_channel": false,
        "view_creator_monetization_analytics": false,
        "view_guild_insights": false,
      },
    },
    {
      "comment": "@everyone",
      "id": "1094801227080015882",
      "permissions": {
        "add_reactions": false,
        "administrator": false,
        "attach_files": false,
        "ban_members": false,
        "change_nickname": false,
        "connect": false,
        "create_instant_invite": false,
        "create_private_threads": false,
        "create_public_threads": false,
        "deafen_members": false,
        "embed_links": false,
        "kick_members": false,
        "manage_channels": false,
        "manage_events": false,
        "manage_guild": false,
        "manage_guild_expressions": false,
        "manage_messages": false,
        "manage_nicknames": false,
        "manage_roles": false,
        "manage_threads": false,
        "manage_webhooks": false,
        "mention_everyone": false,
        "moderate_members": true,
        "move_members": false,
        "mute_members": false,
        "priority_speaker": false,
        "read_message_history": false,
        "request_to_speak": false,
        "send_messages": false,
        "send_messages_in_threads": false,
        "send_tts_messages": false,
        "speak": false,
        "stream": false,
        "use_application_commands": false,
        "use_embedded_activities": false,
        "use_external_emojis": false,
        "use_external_stickers": false,
        "use_soundboard": false,
        "use_vad": false,
        "view_audit_log": false,
        "view_channel": false,
        "view_creator_monetization_analytics": false,
        "view_guild_insights": false,
      },
    },
  ],
  "guildId": "1094801227080015882",
}
`;

exports[`Integrations runs minimal.lua 1`] = `
{
  "categories": [],
  "globalChannels": [],
  "globalRoles": [],
  "guildId": "1094801227080015882",
}
`;

exports[`Integrations runs roles.lua 1`] = `
{
  "categories": [],
  "globalChannels": [],
  "globalRoles": [
    {
      "comment": "test-role",
      "id": "1094801227080015882",
      "permissions": {
        "add_reactions": false,
        "administrator": false,
        "attach_files": false,
        "ban_members": false,
        "change_nickname": false,
        "connect": false,
        "create_instant_invite": false,
        "create_private_threads": false,
        "create_public_threads": false,
        "deafen_members": false,
        "embed_links": false,
        "kick_members": false,
        "manage_channels": false,
        "manage_events": false,
        "manage_guild": false,
        "manage_guild_expressions": false,
        "manage_messages": false,
        "manage_nicknames": false,
        "manage_roles": false,
        "manage_threads": false,
        "manage_webhooks": false,
        "mention_everyone": false,
        "moderate_members": false,
        "move_members": false,
        "mute_members": false,
        "priority_speaker": false,
        "read_message_history": false,
        "request_to_speak": false,
        "send_messages": false,
        "send_messages_in_threads": false,
        "send_tts_messages": false,
        "speak": false,
        "stream": false,
        "use_application_commands": false,
        "use_embedded_activities": false,
        "use_external_emojis": false,
        "use_external_stickers": false,
        "use_soundboard": false,
        "use_vad": false,
        "view_audit_log": false,
        "view_channel": false,
        "view_creator_monetization_analytics": false,
        "view_guild_insights": false,
      },
    },
    {
      "comment": "another-test-role",
      "id": "1094801227080015882",
      "permissions": {
        "add_reactions": false,
        "administrator": false,
        "attach_files": false,
        "ban_members": false,
        "change_nickname": false,
        "connect": false,
        "create_instant_invite": false,
        "create_private_threads": false,
        "create_public_threads": false,
        "deafen_members": false,
        "embed_links": false,
        "kick_members": false,
        "manage_channels": true,
        "manage_events": false,
        "manage_guild": false,
        "manage_guild_expressions": false,
        "manage_messages": false,
        "manage_nicknames": false,
        "manage_roles": false,
        "manage_threads": false,
        "manage_webhooks": false,
        "mention_everyone": false,
        "moderate_members": false,
        "move_members": false,
        "mute_members": false,
        "priority_speaker": false,
        "read_message_history": false,
        "request_to_speak": false,
        "send_messages": false,
        "send_messages_in_threads": false,
        "send_tts_messages": false,
        "speak": false,
        "stream": false,
        "use_application_commands": false,
        "use_embedded_activities": false,
        "use_external_emojis": false,
        "use_external_stickers": false,
        "use_soundboard": false,
        "use_vad": false,
        "view_audit_log": false,
        "view_channel": false,
        "view_creator_monetization_analytics": false,
        "view_guild_insights": false,
      },
    },
    {
      "comment": "another-another-test-role",
      "id": "1094801227080015882",
      "permissions": {
        "add_reactions": false,
        "administrator": false,
        "attach_files": false,
        "ban_members": false,
        "change_nickname": false,
        "connect": false,
        "create_instant_invite": false,
        "create_private_threads": false,
        "create_public_threads": false,
        "deafen_members": false,
        "embed_links": false,
        "kick_members": false,
        "manage_channels": true,
        "manage_events": false,
        "manage_guild": false,
        "manage_guild_expressions": false,
        "manage_messages": true,
        "manage_nicknames": false,
        "manage_roles": false,
        "manage_threads": false,
        "manage_webhooks": false,
        "mention_everyone": false,
        "moderate_members": false,
        "move_members": false,
        "mute_members": false,
        "priority_speaker": false,
        "read_message_history": false,
        "request_to_speak": false,
        "send_messages": false,
        "send_messages_in_threads": false,
        "send_tts_messages": false,
        "speak": false,
        "stream": false,
        "use_application_commands": false,
        "use_embedded_activities": false,
        "use_external_emojis": false,
        "use_external_stickers": false,
        "use_soundboard": false,
        "use_vad": false,
        "view_audit_log": false,
        "view_channel": false,
        "view_creator_monetization_analytics": false,
        "view_guild_insights": false,
      },
    },
  ],
  "guildId": "1094801227080015882",
}
`;
